packages/access-api/src/routes/root.js

```diff
@@ -18,9 +22,84 @@ export async function postRoot(request, env) {
     },
   })
 
-  const rsp = await server.request({
-    body: new Uint8Array(await request.arrayBuffer()),
+  const body = new Uint8Array(await request.arrayBuffer())
+  const headers = Object.fromEntries(request.headers.entries())
+
+  /** @type {[Server.RequestDecoder, Server.ResponseEncoder]} */
+  // We will use encoder / decoder pair based on the content-type header this
+  // way we do not need to have separate routes.
+  const [decoder, encoder] =
+    headers['content-type'] === 'application/car' ? [CAR, CBOR] : [json, json]
+
+  const invocations = await decoder.decode({
+    body,
     headers: Object.fromEntries(request.headers.entries()),
   })
-  return new Response(rsp.body, { headers: rsp.headers })
+
+  // We block until we can log the UCAN invocation if this fails we return a 500
+  // to the client. That is because in the future we expect that invocations will
+  // be written to a queue first and then processed asynchronously, so if we
+  // fail to queue the invocation we should not handle it.
+  await env.ucanLog.logInvocations(body)
+
+  const results = await Promise.all(
+    invocations.map((invocation) => execute(invocation, server, env))
+  )
+
+  const forks = []
+  const out = []
+
+  for (const receipt of results) {
+    out.push(receipt.data.out.error || receipt.data.out.ok)
+    // we don't await the logReceipt call because we want to return the response
+    // to the client as soon as possible. We will however keep the worker alive
+    // until we have logged the receipt. If logging fails we will log the error
+    // as we have no way to handle it here nor we can rollback the invocation.
+    forks.push(
+      env.ucanLog.logReceipt(receipt).catch((error) => env.log.error(error))
+    )
+  }
+
+  const response = await encoder.encode(out)
+
+  ctx.waitUntil(Promise.all(forks))
+
+  return new Response(response.body, {
+    headers: response.headers,
+  })
+}
+
+/**
+ *
+ * @param {Server.Invocation} invocation
+ * @param {Server.ServerView<*>} server
+ * @param {API.RouteContext} env
+ * @returns {Promise<Required<API.Block<API.Receipt>>>}
+ */
+const execute = async (invocation, server, env) => {
+  /** @type {[Server.Result<*, Server.API.Failure>]} */
+  const [result] = await Server.execute([invocation], server)
+  const out = result?.error ? { error: result } : { ok: result }
+
+  // Create a receipt payload for the invocation conforming to the spec
+  // @see https://github.com/ucan-wg/invocation/#8-receipt
+  const payload = {
+    ran: invocation.cid,
+    out,
+    fx: { fork: [] },
+    meta: {},
```

Comment:
Currently there is no way to identify which invocation capability was the root for this receipt without opening the CAR file. This is not ideal from a stream of such events.

We should be able to filter out the events we want to handle per capability so that we do not need to read all the CARs to get to know about them. For example, let's consider a stream consumer (with batches of 20) that tracks number of `store/add` successful invocations. This consumer would be required to go through all the receipts received, read all the CAR files from the invocations and check the capability. If from these 20 CARs there are no `store/add` operations to address, all this processing was useless. 

Thinking about adding capability from invocation in `meta` as a good way to flag this. We would read CARs mentioned in receipts with capability we care about (and could still validate that capability is really what is given in `meta`).

Thoughts?

Reply:
> Currently there is no way to identify which invocation capability was the root for this receipt without opening the CAR file. This is not ideal from a stream of such events.

It's not that bad...right? It's a local bucket so won't be costly. Can we see how it pans out and change if necessary? I'd rather not rely on something in a `meta` field...

Reply:
...or maybe a better idea - the UCAN log service should validate that the receipt belongs to a stored invocation. It shoudl read the invocation and annotate what it sends to the stream with what you need.

Reply:
> ...or maybe a better idea - the UCAN log service should validate that the receipt belongs to a stored invocation. It shoudl read the invocation and annotate what it sends to the stream with what you need.

I think this should be good yes, I will report back once I try it out

Reply:
> It's not that bad...right? It's a local bucket so won't be costly. Can we see how it pans out and change if necessary? I'd rather not rely on something in a meta field...

Depends, if we could use [Kinesis filters already](https://github.com/web3-storage/w3infra/blob/main/stacks/ucan-invocation-stack.js#L215) that would be a bumper if we could not filter out consumers right away. But your solution above should work well